[SPARK-7899][PYSPARK] Fix Python 3 pyspark/sql/types module conflict

mnazario opened the pull request:
This PR makes the types module in `pyspark/sql/types` work with pylint static analysis by removing the dynamic naming of the `pyspark/sql/_types` module to `pyspark/sql/types`.

Tests are now loaded using `$PYSPARK_DRIVER_PYTHON -m module` rather than `$PYSPARK_DRIVER_PYTHON module.py`. The old method adds the location of `module.py` to `sys.path`, so this change prevents accidental use of relative paths in Python.

mnazbro committed:
Move _types back to types

mnazbro committed:
Change test scripts to use "-m"

By changing to use "-m", the directory which contains the test file
(e.g. python/pyspark/sql/types.py) will not be added to sys.path.

mnazbro committed:
Make doctests consistent with other tests

AmplabJenkins commented:
Can one of the admins verify this patch?

justinuang commented on python/run-tests:
```diff
@@ -57,54 +57,54 @@ function run_test() {
-    run_test "pyspark/rdd.py"
-    run_test "pyspark/context.py"
-    run_test "pyspark/conf.py"
-    PYSPARK_DOC_TEST=1 run_test "pyspark/broadcast.py"
```
Why did you remove the PYSPARK_DOC_TEST env var?

mnazario replied:
I removed PYSPARK_DOC_TEST to make the doctests consistent with the other pyspark tests since I was changing all of the other test names to modules instead of filepaths.

The two doctests now have the appropriate main function which runs the doctests for that module.

justinuang replied:
alright, sounds good!

JoshRosen commented:
Jenkins, this is ok to test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33605/consoleFull) for   PR 6439 at commit [`bb8b04d`](https://github.com/apache/spark/commit/bb8b04db418922ac43d814dc5967e61fa03c99c4).

SparkQA commented:
[Test build #33605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33605/consoleFull) for   PR 6439 at commit [`bb8b04d`](https://github.com/apache/spark/commit/bb8b04db418922ac43d814dc5967e61fa03c99c4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33605/
Test PASSed.

JoshRosen commented:
/cc @davies

davies commented:
This looks good to me, could you also update pyspark/mllib/rand.py to random.py? (and remove the hack).

One downside of this is that we can not run single test suite by spark-submit:
```
$ bin/spark-submit python/pyspark/sql/dataframe.py
```
Instead, we should do:
```
$ PYTHON_PATH=$HOME/python:$HOME/python/lib/py4j-0.8.2.1-src.zip python -m pyspark.sql.dataframe
```

mnazario commented:
I'll remove the hack. I don't really think of the spark-submit change as a downside since it's the proper way it should be run for now.
Once this stuff is moved to an actual test running like nose, then you'll be able to selectively choose which tests need to be run that way.

mnazbro committed:
Remove hack on random.py

There is a hack on random.py which is similar to types.py. This
hack should be removed by the run-tests change from source file to
module.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33683 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33683/consoleFull) for   PR 6439 at commit [`366ef30`](https://github.com/apache/spark/commit/366ef30e0063cb69a4174312845b103bb48ec75f).

SparkQA commented:
[Test build #33683 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33683/consoleFull) for   PR 6439 at commit [`366ef30`](https://github.com/apache/spark/commit/366ef30e0063cb69a4174312845b103bb48ec75f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

davies commented:
LGTM

JoshRosen commented:
@mnazario, since this was motivated by running `pylint` checks, I'd be very curious to know whether pylint uncovered any bugs or dodgy code.  Do you have any results that you can share?

mnazario commented:
We were running `pylint` on our code which was depending on pyspark. The dynamic module naming  confused pylint.
